api/src/routes/protected/chapter.ts

```diff
@@ -0,0 +1,86 @@
+import { type FastifyPluginCallbackTypebox } from '@fastify/type-provider-typebox';
+
+import { ProgressTimestamp, getPoints } from '../../utils/progress';
+import * as schemas from '../../schemas';
+
+/**
+ * Plugin for the chapter submission endpoints.
+ *
+ * @param fastify The Fastify instance.
+ * @param _options Options passed to the plugin via `fastify.register(plugin, options)`.
+ * @param done The callback to signal that the plugin is ready.
+ */
+export const chapterRoutes: FastifyPluginCallbackTypebox = (
+  fastify,
+  _options,
+  done
+) => {
+  fastify.post(
+    '/chapter-completed',
+    {
+      schema: schemas.chapterCompleted,
+      errorHandler(error, request, reply) {
+        if (error.validation) {
+          void reply.code(400);
+          return {
+            type: 'error',
+            message: 'That does not appear to be a valid chapter submission.'
+          };
+        } else {
+          fastify.errorHandler(error, request, reply);
+        }
+      }
+    },
+    async req => {
+      const { id } = req.body;
+      const userId = req.user?.id;
+
+      const user = await fastify.prisma.user.findUniqueOrThrow({
+        where: { id: userId },
+        select: {
+          id: true,
+          completedChapters: true,
+          progressTimestamps: true
+        }
+      });
+
+      const alreadyCompletedChapter = user.completedChapters.find(
+        ({ id: chapterId }) => chapterId === id
+      );
+
+      if (alreadyCompletedChapter) {
+        return {
+          alreadyCompleted: true,
+          completedDate: alreadyCompletedChapter.completedDate,
+          points: getPoints(user.progressTimestamps as ProgressTimestamp[])
+        };
+      }
+
+      const completedChapter = {
+        id,
+        completedDate: Date.now()
+      };
+
+      const newProgressTimestamps = [
+        ...(user.progressTimestamps as ProgressTimestamp[]),
+        completedChapter.completedDate
+      ];
```

Comment:
In my head, completing a chapter/module is akin to claiming a certification. i.e. it's not progress, exactly. A learner only currently makes progress by completing challenges. Unless we're planning to change this, I recommend leaving the timestamps alone for now.

We can always fix this at a later date, since you're storing the id and completion date in `completedChapters`.